### ISSUE TITLE
[Bug] training (sometimes) freezes in a multi-gpu setting without throwing any errors or warnings.

### ISSUE THREAD

**Issue by ragavsachdeva** [PyTorchLightning/pytorch-lightning#11841]
## 🐛 Bug

tldr: Training freezes in a multi-gpu setting without throwing any errors or warnings.

I was trying to train my model on 2 GPUs using the ddp strategy and it would consistently freeze on `Epoch 0, Batch 40`. No errors, no warnings. Almost like a deadlock. I put in some print statements and found out that the `training_step` was never called for `Batch 41` on `process 0` but it was for `process 1`.
Then I ran the same experiment using 1 GPU and noticed that after `Epoch 0, Batch 80`, the trainer just silently moved onto `Epoch 1` without finishing `Epoch 0` or throwing any errors (screenshot below), and interestingly `Epoch 1` executed normally up to 100%.

![Screenshot 2022-02-10 at 10 41 20 am](https://user-images.githubusercontent.com/26804893/153426107-c09ba541-e0e5-4dc3-a2e1-594b7dfd749a.png)

It was only when I used LightningLite with 1 GPU the program crashed at `Epoch 0, Batch 80` leading me to find that there was an index-out-of-bound bug in my augmentation pipeline (which would trigger infrequently). The pytorch-lightning script never crashed.

Below is a script to reproduce the freeze/deadlock using the BoringModel script. You might need to run it a few times to reproduce the freeze (it does catch the list out of bound error sometimes). Here is the screenshot of the output produced by the script below, where it is frozen at `Epoch 2, Batch 1`.

![Screenshot 2022-02-10 at 2 29 13 pm](https://user-images.githubusercontent.com/26804893/153428241-f0c95e38-e583-46a5-aa9c-9a90aef37123.png)

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
import torch.nn as nn
from random import random

class AugmentationPipeline(nn.Module):
    def __init__(self) -> None:
        super(AugmentationPipeline, self).__init__()
    
    def cause_some_error(self):
        array = []
        invalid_index = array[0]

    def forward(self, x):
        if random() < 0.02:
            self.cause_some_error()
        return x

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)
        self.augmentations = AugmentationPipeline()

    def __getitem__(self, index):
        return self.augmentations(self.data[index])

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        print("...")
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        gpus=2,
        strategy="ddp",
        default_root_dir=os.getcwd(),
        limit_train_batches=2,
        limit_val_batches=2,
        limit_test_batches=2,
        num_sanity_val_steps=0,
        max_epochs=3,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()
```

### Expected behavior

I expect the program to crash with trace rather than freeze.

### Environment

```
* CUDA:
	- GPU:
		- NVIDIA GeForce GTX TITAN X
		- NVIDIA GeForce GTX TITAN X
		- NVIDIA GeForce GTX TITAN X
		- NVIDIA GeForce GTX TITAN X
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.21.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.10.1
	- pytorch-lightning: 1.5.8
	- tqdm:              4.62.3
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.9
	- version:           #1 SMP Fri Jan 14 13:59:45 UTC 2022
```

### Additional context

Note: The `random()` in the script above is important to reproduce the behaviour.

cc @justusschock @kaushikb11 @awaelchli @akihironitta @rohitgr7

**Comment by awaelchli** [comment]
Hello
Yes I agree what you are seeing makes sense and unfortunately Lightning can't show the error from the subprocesses. Definitely a pain point, I agree. 

We have a deadlock detection mechanism in Lightning that can be activated by setting the environment variable `PL_RECONCILE_PROCESS=1`. This is an experimental feature but we should mention it in our debugging section at least to give it more visibility.

**Comment by ragavsachdeva** [comment]
@awaelchli thanks for the update!

Let's ignore multi-gpu setting for now. When I set `--gpus=1`, Pytorch Lightning just silently dropped the epoch and moved onto the next when the augmentation pipeline failed (see the 1st screenshot in my post). However, LightningLite correctly crashed. Can you help me understand the discrepancy in this behaviour? Not being able to __see__ the error message is definitely a pain point but I'd still expect the program to crash.

**Comment by awaelchli** [comment]
I can reproduce it with your script on GPU, pytorch 1.10, with lightning 1.5.8, 1.5.9, 1.5.10 but NOT on the latest commit on master.

On master, I get this error consistently (I ran it 15+ times) and the script does not hang.
I'm not sure exactly what might have fixed it. 

```
Epoch 0:  25%|██▌       | 1/4 [00:00<00:00,  9.07it/s, loss=2.68, v_num=39]...
...
Epoch 0:  50%|█████     | 2/4 [00:00<00:00, 17.49it/s, loss=2.35, v_num=39]
Validating: 0it [00:00, ?it/s]
Validating:   0%|          | 0/2 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 685, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 777, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 1199, in _run
    self._dispatch()
  

[ TRUNCATED ]

  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 32, in __getitem__
    return self.augmentations(self.data[index])
  File "/Users/adrian/.conda/envs/lightning/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 21, in forward
    self.cause_some_error()
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 17, in cause_some_error
    invalid_index = array[0]
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 87, in <module>
    run()
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 82, in run
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 740, in fit
    self._call_and_handle_interrupt(
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/trainer/trainer.py", line 698, in _call_and_handle_interrupt
    self.training_type_plugin.reconciliate_processes(traceback.format_exc())
  File "/Users/adrian/repositories/pytorch-lightning/pytorch_lightning/plugins/training_type/ddp.py", line 533, in reconciliate_processes
    raise DeadlockDetectedException(f"DeadLock detected from rank: {self.global_rank} \n {trace}")


[ TRUNCATED ]


  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 32, in __getitem__
    return self.augmentations(self.data[index])
  File "/Users/adrian/.conda/envs/lightning/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 21, in forward
    self.cause_some_error()
  File "/Users/adrian/repositories/pytorch-lightning/pl_examples/bug_report/bug_report_model.py", line 17, in cause_some_error
    invalid_index = array[0]
IndexError: list index out of range
```

**Comment by Martin0Li** [comment]
I have a similar issue where the training process gets stuck after one epoch when using multiple GPUs and using val_dataloaders in trainer.fit.
Looking forward to your reply, thank you. @awaelchli

**Comment by AugustDev** [comment]
Any ideas, I still get freezes?

**Comment by Neupanesujal** [comment]
If you are training video classification then if the total frame in that video is 0 then model may gets freeze while training.
